Improve is_hermitian property for class Mul

mathiscr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Related to #27989 

#### Brief description of what is fixed or changed
This PR improves the `is_hermitian` property for the Mul class by adding a test based on the definition of a Hermitian object. It checks if A.adjoint() == A.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * Added a test in the property _eval_is_hermitian for the Mul class that check the definition of a Hermitian object. For example, now, `A*Dagger(A).is_hermitian is True` because `Dagger(A*Dagger(A)) == A*Dagger(A)`
<!-- END RELEASE NOTES -->

mathiscr committed:
Improve is_hermitian property for class Mul

This commit improves the is_hermitian property for the Mul class by adding a test based on the definition of a Hermitian object.

akhmerov commented on sympy/core/mul.py:
```diff
@@ -1541,9 +1541,15 @@ def _eval_is_imaginary(self):
 
     def _eval_is_hermitian(self):
+        from sympy.simplify import simplify
+        if simplify(self.func(*self.args) - self._eval_adjoint()).is_zero:
```
`simplify` has potentially very long evaluation time, I am not sure this is how other similar methods behave, rather than failing early.

mathiscr replied:
I used simplify because simplify is used here:
https://github.com/sympy/sympy/blob/a6f35dbfa24ba66ff8f664a7fae7011a435f0832/sympy/matrices/matrixbase.py#L1331-L1333
I think that `simp_func` refers to `simplify`, but maybe these two methods are not that similar.

akhmerov replied:
Digging deeper, I became quite sure that automatic simplification is undesirable here. Consider that `sympy.Expr._eval_transpose` contains a check for `is_hermitian`. That would mean that transposing a matrix with a complex expression containing complicated expressions with noncommutative entries will effectively stop working.

oscarbenjamin replied:
The idea with that code is that a user can pass in their own function for simplifying things. It is not a good model to base other things on.

Generally simplify should be avoided within the sympy codebase because it is both slow and heuristic making things not well defined. Usually the suggestion instead is either to use a more targeted simplification function or just not to use any simplification.

akhmerov replied:
I checked the code for matrix property testing. All public methods *ask* the user whether to simplify in addition. These methods also aren't called by sympy internally, as indicated by the comment right before the code you quoted.

TiloRC replied:
`self.func(*self.args) - self._eval_adjoint()` automatically simplifies to zero — at least in the cases currently being tested. I'm not sure what cases if any were being handled because of the call to simplify. If you had some cases that could only be handled with the call to simplify in mind, they should probably be added as tests.

mathiscr replied:
> self.func(*self.args) - self._eval_adjoint() automatically simplifies to zero

I agree with you, I think that simplify was not needed here.

TiloRC committed:
Remove call to simplify and simplify code

TiloRC commented:
You should have tests for `is_antihermitian`. Also, it would be nice if we returned False when the call to `.is_zero` came back false. `.is_zero` can also return None and you have to be careful to treat that differently than false.

akhmerov commented on sympy/core/mul.py:
```diff
@@ -1541,9 +1541,13 @@ def _eval_is_imaginary(self):
             return self._eval_real_imag(False)
 
     def _eval_is_hermitian(self):
+        if (self - self._eval_adjoint()).is_zero:
```
I believe this still has the problem I described in the following comment: https://github.com/sympy/sympy/pull/27989#issuecomment-2827302481

While simplifying is not what we want `(self - self._eval_adjoint()).is_zero` will fail on even very simple expressions.

TiloRC replied:
That issue is unfortunate. I believe `(self - self._eval_adjoint()) == 0` passes the existing tests, so maybe we could just do that for now.

akhmerov replied:
Perhaps an incrementally better option to consider is `self == self.adjoint()`? That would cover the case of `Dagger(A) * A`

mathiscr replied:
Yes, both `self - self._eval_adjoint() == 0` and `self = self.adjoint()` passe the existing tests. Both implementation works for me.

mathiscr commented:
I think that this test doesn't pass:
```python
m = Matrix([[0, I], [-I, 0]])
id = eye(2)
n = MatMul(m, id, evaluate=False)
print(n.is_hermitian) # False
```

Maybe it is because `MatMul` inherits the `_eval_is_hermitian()` method from `Mul`, and this does not work properly for `MatMul`.

oscarbenjamin commented:
> both self - self._eval_adjoint() == 0 and self = self.adjoint()

Both of these seem backwards to me. There as queries like `is_real`, `is_hermitian` etc and then there is symbolic evaluation like `sqrt(x)**2 -> x`. Much logic in sympy is tide up in the evaluation rules so people implementing queries like `is_real` are often tempted to use evaluation as a way of resolving the queries. It should be the other way round though: the rules for evaluation should depend on the queries. Rather than `is_hermitian` calling `adjoint` it should be that `adjoint` can check `is_hermitian`.

Essentially what you are trying to do is implement a low-level things using a high-level thing. Rather than looking around at what is already there the proper things is just to write the code that implements the query. If `.adjoint()` can do this then it can also be implemented directly in `is_hermitian`. The query can be resolved by writing some code that analyses the structure of the expression rather than by creating new expressions.

TiloRC committed:
Add test cases and make other test cases match style

TiloRC commented:
> The query can be resolved by writing some code that analyses the structure of the expression rather than by creating new expressions.

I think we need to go back to the code that checks if the Mul arguments are "palindrome-like" with respect to comparing whether two elements are conjugates of each other. We should be using a pattern matching approach, so we can't use the previous strategies we've used to check if two elements are adjoint of each other. Instead, I think we should do something like this:

```py
def _is_adjoint_pair(expr1, expr2):
  if isinstance(expr1, Adjoint):
    if expr1.args[0] == expr2:
      return True
  if isinstance(expr2, Adjoint):
    if expr2.args[0] == expr1:
      return True

  if expr1.is_hermitian and expr1 == expr2:
    return True
```

Something a bit tricky to deal with are scalars. When doing the "palindrome-like" structural comparison, scalars need to be ignored. However, we still need to check whether the scalar is hermitian — something like `(xnc.adjoint()*(I+10)*xnc)` is not hermitian.

mathiscr commented:
Using `isinstance( . , Adjoint)` could be problematic, because when I run
```python
x = symbols("x", commutative=False)
print(type(x.adjoint())
```
I get `Dagger` which is defined in the physics module.

mathiscr commented:
```python
    def _eval_is_hermitian(self):
        from sympy import adjoint

        def _is_adjoint_pair(expr1, expr2):
            if isinstance(expr1, adjoint) and expr1.args[0] == expr2:
                return True
            if isinstance(expr2, adjoint) and expr2.args[0] == expr1:
                return True
            return fuzzy_and((
                fuzzy_or((expr1.is_hermitian, expr1.is_antihermitian)),
                expr1 == expr2
            ))

        # return True if the product of the list is real
        def is_list_real(l):
            return fuzzy_and((x.is_real for x in l))    # Could be improved because [i, i] is True

        # return True if the product of the list is imaginary
        def is_list_imaginary(l):
            n_imaginary = 0
            for z in l:
                if z.is_imaginary is True:
                    n_imaginary += 1
                elif fuzzy_not(fuzzy_or((z.is_imaginary, z.is_real))):
                    return None
            return (n_imaginary % 2 == 1)

        c_part, nc_part = self.args_cnc()
        n = len(nc_part)

        if n == 0:
            return self._eval_herm_antiherm(True)

        all_pairs_adjoint = fuzzy_and(
            _is_adjoint_pair(nc_part[i], nc_part[n - i - 1])
            for i in range(n // 2)
        )

        if all_pairs_adjoint:
            if n % 2 == 0 and is_list_real(c_part):
                return True
            else:
                middle = nc_part[n//2]
                is_middle_hermitian = fuzzy_or((
                    fuzzy_and((middle.is_hermitian, is_list_real(nc_part))),
                    fuzzy_and((middle.is_antihermitian, is_list_imaginary(nc_part)))
                ))
                if is_middle_hermitian:
                    return True

        return self._eval_herm_antiherm(True)
```

I tried to use `_is_adjoint_pair()` in `_eval_is_hermitian()`. I improved it slightly to handle cases where we have two anti-Hermitian matrices (since if `A` is anti-Hermitian, then `_is_adjoint_pair(A, A)` should return `True`).
I don't think this code creates any new expressions that would violate the "don't create expressions" rule, but maybe I'm wrong.

TiloRC commented:
> I don't think this code creates any new expressions that would violate the "don't create expressions" rule, but maybe I'm wrong.

I agree.

---

I won't give any feedback to the code yet. You should add it to the actual code and push your changes. Also, you should add a lot more test cases that handle. Aim for full coverage — for more info on writing tests see https://docs.sympy.org/latest/contributing/new-contributors-guide/writing-tests.html.

mathiscr committed:
Improve _eval_is_hermitian _eval_is_antihermitian

This commit will improve _eval_is_hermitian and _eval_is_antihermitian.
With this improvement, _eval_is_hermitian will check if an expression is hermitian by
checking the definition of an hermitian object ie A.adjoint() = A.
To do that, we use a pattern matching approach to avoid the creation of
new expressions. We will check if the expression is "Palindrome-like"
with adjoint.

mathiscr committed:
Add test cases

mathiscr committed:
correction of _eval_is_antihermitian, more tests

TiloRC commented on sympy/core/mul.py:
```diff
@@ -1541,9 +1540,118 @@ def _eval_is_imaginary(self):
             return self._eval_real_imag(False)
 
     def _eval_is_hermitian(self):
+        from sympy import adjoint
```
You should import it like this: from sympy.functions.elementary.complexes import adjoint

TiloRC replied:
Maybe you'll be able to do this import at the top of the file if you do it that way.

TiloRC replied:
I also think there might be multiple different adjoints in SymPy, so it might be good to check if the current code works for matrices and operators as I suspect their adjoints may be different.

mathiscr replied:
When I try to use `from sympy.functions.elementary.complexes import adjoint` at the beginning of the file, I get a circular import error. Should I import it inside the function, or should I try another way to resolve this problem?

I think there is also an adjoint function for matrices, but there is also a `Mul` class for general expressions and a `MatMul` class specifically for matrices. So, should we add the check for the `adjoint` of matrices inside the `Mul` class or should we create another `_eval_is_hermitian` inside the `MatMul` class with the adjoint of the matrices ?

TiloRC replied:
> Should I import it inside the function, or should I try another way to resolve this problem?

I think it's fine to import inside the function—just make sure you import it with `from sympy.functions.elementary.complexes import adjoint` rather than just `from sympy import adjoint`. There's other SymPy code that takes this approach. I'm not sure if there's some better way to do it. 

> MatMul

MatMul inherits from the Mul class so it's possible they use the same logic. Although, I imagine MatMul probably overrides the logic. It might make sense to create another `_eval_is_hermitian` method if one doesn't already exist. Though, if we do that we should avoid duplicating logic that would be shared by both methods somehow.

TiloRC commented on sympy/core/mul.py:
```diff
@@ -1541,9 +1540,118 @@ def _eval_is_imaginary(self):
+                return True
+            if isinstance(expr2, adjoint) and expr2.args[0] == expr1:
+                return True
+            return fuzzy_and((
+                fuzzy_or((expr1.is_hermitian, expr1.is_antihermitian)),
+                expr1 == expr2
+            ))
```
```py
fuzzy_and((
fuzzy_or((expr1.is_hermitian, expr1.is_antihermitian)),
expr1 == expr2
))
```
will give False if `expr1 != expr2` but I don't think that's correct. Consider
```py
Mul(xnc, xnc, evaluate=False) == xnc*xnc # gives False
```
so `_is_adjoint_pair(Mul(xnc, xnc, evaluate=False), (xnc*xnc).adjoint())` would wrongly give False.

TiloRC replied:
Also, if `expr1.is_antihermitian` and `expr1 == expr2` it's not the case that expr1 is the adjoint of expr2 as far as I'm aware — this would be the case if `expr1 == -expr2`.

mathiscr replied:
> _is_adjoint_pair(Mul(xnc, xnc, evaluate=False), (xnc*xnc).adjoint()) would wrongly give False

Yes, that's a case I hadn't considered, and I don't know how to handle it properly without relying on high level functions like `simplify`.

> Also, if expr1.is_antihermitian and expr1 == expr2 it's not the case that expr1 is the adjoint of expr2 as far as I'm aware — this would be the case if expr1 == -expr2

That's true. However, the `.is_antihermitian` check isn't meant to verify if A and B are adjoints of each other. It's just a particular case that happens to work. So the function name `is_adjoint_pair` might be misleading here. We're checking for `.is_antihermitian` and `expr1 == expr2` because if $A$ is anti-hermitian, then:



$` \begin{align}
adjoint(A*u*A) &= adjoint(A)*adjoint(u)*adjoint(A) \\
                            &= (-A)*adjoint(u)*(-A) \\ 
                            &= A*adjoint(u)*A 
\end{align} `$

So if $u = adjoint(u)$ then $` adjoint(A*u*A) = A*u*A  `$ but maybe this check could lead to wrong result because the reverse implication isn't always true.

TiloRC replied:
> Yes, that's a case I hadn't considered, and I don't know how to handle it properly without relying on high level functions like simplify.

It's alright if we don't handle every case. If `.is_hermitian` sometimes returns None for cases are definitely hermitian or not hermitian that's okay.

TiloRC replied:
>  the .is_antihermitian check isn't meant to verify if A and B are adjoints of each other

Oh, I see the logic. That makes sense. We should probably rename the function and add a comment explaining this though. 

> maybe this check could lead to wrong result because the reverse implication isn't always true

Why would this lead to wrong results?

mathiscr replied:
> Why would this lead to wrong results?

Because the equivalence is not True, we only have : $` u = adjoint(u) \Rightarrow A*u*A = A*adjoint(u)*A `$ but this implication could be sufficient for our purposes.

TiloRC replied:
I think we should replace the current code with the following:
```python
if expr1 == expr2 and (expr1.is_hermitian or expr1.is_antihermitian):
    return True 
```

This code is simpler and also doesn't wrongly return False sometimes.

TiloRC commented on sympy/core/mul.py:
```diff
@@ -1541,9 +1540,118 @@ def _eval_is_imaginary(self):
+            for z in l:
+                if z.is_imaginary is True:
+                    n_imaginary += 1
+                elif fuzzy_not(fuzzy_or((z.is_imaginary, z.is_real))):
+                    return None
```
If `fuzzy_not` gives None, None should be returned here. Because it's not, we get wrong behavior like this:
```py
from sympy.core.logic import fuzzy_not, fuzzy_and, fuzzy_or
def is_list_imaginary(l):
    n_imaginary = 0
    for z in l:
        if z.is_imaginary is True:
            n_imaginary += 1
        elif fuzzy_not(fuzzy_or((z.is_imaginary, z.is_real))):
            return None
    return (n_imaginary % 2 == 1)

from sympy.abc import x

is_list_imaginary([x]) # wrongly gives False
```

TiloRC replied:
By the way, it would be good to add more tests for each "bug" I've noticed in the current code.

TiloRC commented on sympy/core/mul.py:
```diff
@@ -1541,9 +1540,118 @@ def _eval_is_imaginary(self):
+        if n == 0:
+            return self._eval_herm_antiherm(True)
+
+        all_pairs_adjoint = fuzzy_and(
+            _is_adjoint_pair(nc_part[i], nc_part[n - i - 1])
+            for i in range(n // 2)
+        )
+        if all_pairs_adjoint:
+            if n % 2 == 0 :
+                if is_list_real(c_part):
+                    return True
+            else:
+                middle = nc_part[n//2]
+                is_middle_hermitian = fuzzy_or((
+                    fuzzy_and((middle.is_hermitian, is_list_real(c_part))),
+                    fuzzy_and((middle.is_antihermitian, is_list_imaginary(c_part)))
+                ))
+                if is_middle_hermitian:
+                    return True
```
This code is more complicated than it needs to be. Rather than handling the even and odd case separately, I think it would make more sense to go up to `(n+1)//2` in `range(n // 2)` in the following code:
```py
all_pairs_adjoint = fuzzy_and(
    _is_adjoint_pair(nc_part[i], nc_part[n - i - 1])
    for i in range(n // 2)
)
```

mathiscr replied:
I think we can't go up to `(n+1)//2` with the current implementation of `is_pair_adjoint`, because if we evaluate `is_adjoint_pair(middle, middle)` where middle is hermitian and if we have an imaginary scalar in front of the main expression, then `_eval_is_hermitian` will return True. This happens because `is_adjoint_pair(middle, middle)` returns True, but the expression is actually not hermitian the imaginary scalar changes sign under the adjoint operation, so the overall expression is not equal to its adjoint. So I think that we have to consider the odd case as a special case but maybe I'm wrong.

TiloRC replied:
> but maybe I'm wrong

You're saying that if the middle was antihermitian, then `is_pair_adjoint(middle, middle)` would give True which would result in a wrong result. That makes sense.

Regardless, I do still think that the code is more complicated than it needs to be. I feel like there should be some ways to improve it.

mathiscr replied:
> You're saying that if the middle was antihermitian, then is_pair_adjoint(middle, middle) would give True which would result in a wrong result?

Yes, it would also doesn't work if the middle is antihermitian and if the scalar in front of the expression is not an imaginary number. For exemple, if $`A`$ is antihermitian, then the middle ($`A`$) is antihermitian so `is_pair_adjoint` would return `True` but the expression is not hermitian. But if the scalar in front of the expression is imaginary, then the expression would be hermitian ($`iA`$ is hermitian). 
So I think that when we check if the middle is hermitian or antihermitian, we should also check the scalar to handle the previous case : if the middle is hermitian and the scalar is real, then the expression is hermitian (if A is hermitian then 2A is hermitian but iA is not hermitian) and if the middle is antihermitian and the scalar is an imaginary number, then the expression is hermitian (if A is antihermitian then 2A is not hermitian but iA is hermitian).
But maybe it is possible to simplify this logic to simply the code.

mathiscr committed:
Fix some bugs

mathiscr committed:
Add a comment to explain a function

TiloRC commented on sympy/core/mul.py:
```diff
@@ -1541,9 +1540,120 @@ def _eval_is_imaginary(self):
     def _eval_is_hermitian(self):
+        from sympy.functions.elementary.complexes import adjoint
+
+        def _is_adjoint_pair(expr1, expr2):
```
Maybe _is_adjoint_multiple would be a better name? And then add a comment explaining what it does.

mathiscr committed:
Add comments and simplify the logic

TiloRC commented on sympy/core/mul.py:
```diff
@@ -1541,9 +1540,111 @@ def _eval_is_imaginary(self):
+                    return True
+            else:
+                middle = nc_part[n//2]
+                is_middle_hermitian = fuzzy_or((
+                    fuzzy_and((middle.is_hermitian, get_type(c_part))),
+                    fuzzy_and((middle.is_antihermitian, fuzzy_not(get_type(c_part))))
+                ))
```
again, I think using fuzzy_or / fuzzy_and is overcomplicating things. Is there a reason why we would want to differentiate between None / False here?
